## Author: Birju Dhaduk
## Date: 20200330
## CS301-006, Professor Watson
## HW06 Solution

The purpose of this assignment is to help students practice using Python to prepare their data sets for machine learning.

link to the git repo:

In [96]:
import pandas as pd
import numpy as np

#### Question 1 - Read in the data set and print the % empty values for each column. Are there any columns with missing values?

In [97]:
df = pd.read_csv('C:\\Users\\birju\\Documents\\CS 301\\groundhog.csv')
df

Unnamed: 0  Year Punxsutawney Phil  February Average Temperature  \
0             0  1886         No Record                           NaN   
1             1  1887       Full Shadow                           NaN   
2             2  1888       Full Shadow                           NaN   
3             3  1889         No Record                           NaN   
4             4  1890         No Shadow                           NaN   
..          ...   ...               ...                           ...   
126         126  2012       Full Shadow                         37.51   
127         127  2013         No Shadow                         34.77   
128         128  2014       Full Shadow                         32.13   
129         129  2015       Full Shadow                         32.99   
130         130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                            NaN                        NaN   
1                                            NaN                        NaN   
2                                            NaN                        NaN   
3                                            NaN                        NaN   
4                                            NaN                        NaN   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                             

In [98]:
def  get_percent_nans(df, column_name):
    colSize = df.shape[0]
    empty = df.loc[df[column_name].isna()]
    emptySize = empty.shape[0]
    return (emptySize/colSize) * 100

In [99]:
for col in df.columns:
    print('The \"{}\" column has {:.2f}% empty values'.format(col, get_percent_nans(df, col)))

The "Unnamed: 0" column has 0.00% empty values
The "Year" column has 0.00% empty values
The "Punxsutawney Phil" column has 0.00% empty values
The "February Average Temperature" column has 6.87% empty values
The "February Average Temperature (Northeast)" column has 6.87% empty values
The "February Average Temperature (Midwest)" column has 6.87% empty values
The "February Average Temperature (Pennsylvania)" column has 6.87% empty values
The "March Average Temperature" column has 6.87% empty values
The "March Average Temperature (Northeast)" column has 6.87% empty values
The "March Average Temperature (Midwest)" column has 6.87% empty values
The "March Average Temperature (Pennsylvania)" column has 6.87% empty values


There are 8 columns with 9 (6.87%) missing values in each corresponding to the same 9 rows

#### Question 2 - Create a new dataframe which is a copy of the orginal one. In this copy, replace the missing values in the data set using one of the discussed methods in class.

In [100]:
cp_df = df.copy()
columns = ['February Average Temperature', 'February Average Temperature (Northeast)', 'February Average Temperature (Midwest)', 'February Average Temperature (Pennsylvania)', 'March Average Temperature', 'March Average Temperature (Northeast)', 'March Average Temperature (Midwest)', 'March Average Temperature (Pennsylvania)']
for col in columns:
    for i in range(0,9):
        if cp_df.loc[i, 'Punxsutawney Phil'] == 'No Record':
            cp_df.loc[i, col] = (cp_df.loc[(cp_df['Punxsutawney Phil'] == 'No Record')])[col].median()
        if cp_df.loc[i, 'Punxsutawney Phil'] == 'No Shadow':
            cp_df.loc[i, col] = (cp_df.loc[(cp_df['Punxsutawney Phil'] == 'No Shadow')])[col].median()
        if cp_df.loc[i, 'Punxsutawney Phil'] == 'Partial Shadow':
            cp_df.loc[i, col] = (cp_df.loc[(cp_df['Punxsutawney Phil'] == 'Partial Shadow')])[col].median()
        if cp_df.loc[i, 'Punxsutawney Phil'] == 'Full Shadow':
            cp_df.loc[i, col] = (cp_df.loc[(cp_df['Punxsutawney Phil'] == 'Full Shadow')])[col].median()

In [101]:
for col in cp_df.columns:
    print('The \"{}\" column has {:.2f}% empty values'.format(col, get_percent_nans(cp_df, col)))

The "Unnamed: 0" column has 0.00% empty values
The "Year" column has 0.00% empty values
The "Punxsutawney Phil" column has 0.00% empty values
The "February Average Temperature" column has 0.00% empty values
The "February Average Temperature (Northeast)" column has 0.00% empty values
The "February Average Temperature (Midwest)" column has 0.00% empty values
The "February Average Temperature (Pennsylvania)" column has 0.00% empty values
The "March Average Temperature" column has 0.00% empty values
The "March Average Temperature (Northeast)" column has 0.00% empty values
The "March Average Temperature (Midwest)" column has 0.00% empty values
The "March Average Temperature (Pennsylvania)" column has 0.00% empty values


In [102]:
cp_df

Unnamed: 0  Year Punxsutawney Phil  February Average Temperature  \
0             0  1886         No Record                        32.425   
1             1  1887       Full Shadow                        33.575   
2             2  1888       Full Shadow                        33.575   
3             3  1889         No Record                        32.425   
4             4  1890         No Shadow                        35.550   
..          ...   ...               ...                           ...   
126         126  2012       Full Shadow                        37.510   
127         127  2013         No Shadow                        34.770   
128         128  2014       Full Shadow                        32.130   
129         129  2015       Full Shadow                        32.990   
130         130  2016         No Shadow                        39.470   

     February Average Temperature (Northeast)  \
0                                       21.15   
1                                       22.95   
2                                       22.95   
3                                       21.15   
4                                       24.10   
..                                        ...   
126                                     30.60   
127                                     25.30   
128                                     21.00   
129                                     12.10   
130                                     28.10   

     February Average Temperature (Midwest)  \
0                                     28.55   
1                                     32.95   
2                                     32.95   
3                                     28.55   
4                                     34.30   
..                                      ...   
126                                   38.60   
127                                   33.40   
128                                   27.10   
129                                   23.40   
130                                   36.90   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                          23.80                     38.645   
1                                          26.95                     41.810   
2                                          26.95                     41.810   
3                                          23.80                     38.645   
4                                          27.40                     42.800   
..                                           ...                        ...   
126                                        34.30                     50.410   
127                                        27.40                     40.910   
128                                        22.80                     40.510   
129                                        16.10                     45.390   
130                                        30.80                     47.500   

     March Average Temperature (Northeast)  \
0                                     29.9   
1                                     32.4   
2                                     32.4   
3                                     29.9   
4                                     33.0   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
0                                  39.35   
1                                  42.70   
2                                  42.70   
3                                  39.35   
4                                  43.50   
..                                   ...   
126                                56.30   
127                                37.30   
128                                38.00   
129                             

#### Question 3 - Is your proposed method from Question 2 appropriate in this case? Why or why not?

The method I used in Question 2 is replacing the empty values with the median value of each respective column from the rows that have the same value in the 'Punxsutawney Phil' column.

Although this is the best that could be done given the information given, this method is not appropriate as the rows that have "No Record" could actually be any of the other 3 options but we can not know which one and thus making predictions for those rows is not appropriate. It also does not make sense to make predictions for 8 rows when there are only values given in 3 of the rows.

#### Question 4 - Using the original dataframe (unchanged from Questoins 2, 3), create a mapping for the "Punxsutawney Phil" column which changes the categories into numerical quantities. For example: 'No Record' can be 0, 'No Shadow' can be 1, 'Partial Shadow' can be 2, and 'Full Shadow' can be 3. You should create the mapping *and* apply it to the data set by adding a new column called "Shadow".

In [103]:
for i in range(0,131):
    if df.loc[i, 'Punxsutawney Phil'] == 'No Record':
        df.loc[i, 'Shadow'] = 0
    if df.loc[i, 'Punxsutawney Phil'] == 'No Shadow':
        df.loc[i, 'Shadow'] = 1
    if df.loc[i, 'Punxsutawney Phil'] == 'Partial Shadow':
        df.loc[i, 'Shadow'] = 2
    if df.loc[i, 'Punxsutawney Phil'] == 'Full Shadow':
        df.loc[i, 'Shadow'] = 3
df

Unnamed: 0  Year Punxsutawney Phil  February Average Temperature  \
0             0  1886         No Record                           NaN   
1             1  1887       Full Shadow                           NaN   
2             2  1888       Full Shadow                           NaN   
3             3  1889         No Record                           NaN   
4             4  1890         No Shadow                           NaN   
..          ...   ...               ...                           ...   
126         126  2012       Full Shadow                         37.51   
127         127  2013         No Shadow                         34.77   
128         128  2014       Full Shadow                         32.13   
129         129  2015       Full Shadow                         32.99   
130         130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
0                                         NaN   
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                            NaN                        NaN   
1                                            NaN                        NaN   
2                                            NaN                        NaN   
3                                            NaN                        NaN   
4                                            NaN                        NaN   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                             

#### Question 5 - Remove the rows which have empty values in all of the quantitative columns. (Only 9 should be removed.)

In [104]:
drop = range(0, 9)
df = df.drop(df.index[drop])

In [108]:
df

Unnamed: 0  Year Punxsutawney Phil  February Average Temperature  \
9             9  1895         No Record                         26.60   
10           10  1896         No Record                         35.04   
11           11  1897         No Record                         33.39   
12           12  1898       Full Shadow                         35.37   
13           13  1899         No Record                         25.50   
..          ...   ...               ...                           ...   
126         126  2012       Full Shadow                         37.51   
127         127  2013         No Shadow                         34.77   
128         128  2014       Full Shadow                         32.13   
129         129  2015       Full Shadow                         32.99   
130         130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
9                                        15.6   
10                                       22.2   
11                                       23.6   
12                                       24.8   
13                                       18.1   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
9                                      21.9   
10                                     33.5   
11                                     34.7   
12                                     33.3   
13                                     22.2   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
9                                           17.0                      39.97   
10                                          26.6                      38.03   
11                                          27.9                      38.79   
12                                          26.7                      41.05   
13                                          20.0                      37.63   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
9                                     27.6   
10                                    25.3   
11                                    32.0   
12                                    38.0   
13                                    29.3   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
9                                   40.2   
10                                  36.9   
11                                  44.0   
12                                  46.0   
13                                  38.4   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                             

#### Question 6 - For the remaining rows which have no record (but the other columns *are* populated), determine an appropriate method for inputting the value for this column. Explain your reasoning, and show your work below (using pandas).

In [127]:
no_shadow = df.loc[df['Shadow'] == 1]
partial_shadow = df.loc[df['Shadow'] == 2]
full_shadow = df.loc[df['Shadow'] == 3]

stats = ['No Median', 'No Mean', 'No StD', 'Partial Median', 'Partial Mean', 'Partial StD', 'Full Median', 'Full Mean', 'Full StD']
d = {'Stats':stats}
stats_df = pd.DataFrame(d)

columns = ['Year', 'February Average Temperature', 'February Average Temperature (Northeast)', 'February Average Temperature (Midwest)', 'February Average Temperature (Pennsylvania)', 'March Average Temperature', 'March Average Temperature (Northeast)', 'March Average Temperature (Midwest)', 'March Average Temperature (Pennsylvania)']
for col in columns:
    stats_df.loc[0, col] = no_shadow[col].median()
    stats_df.loc[1, col] = no_shadow[col].mean()
    stats_df.loc[2, col] = no_shadow[col].std()
    stats_df.loc[3, col] = partial_shadow[col].median()
    stats_df.loc[4, col] = partial_shadow[col].mean()
    stats_df.loc[5, col] = partial_shadow[col].std()
    stats_df.loc[6, col] = full_shadow[col].median()
    stats_df.loc[7, col] = full_shadow[col].mean()
    stats_df.loc[8, col] = full_shadow[col].std()
stats_df

Stats         Year  February Average Temperature  \
0       No Median  1990.000000                     35.550000   
1         No Mean  1987.600000                     35.578000   
2          No StD    23.042817                      2.189221   
3  Partial Median  1942.000000                     30.740000   
4    Partial Mean  1942.000000                     30.740000   
5     Partial StD          NaN                           NaN   
6     Full Median  1953.500000                     33.575000   
7       Full Mean  1953.830000                     33.712500   
8        Full StD    33.503522                      3.249446   

   February Average Temperature (Northeast)  \
0                                 24.100000   
1                                 23.486667   
2                                  4.767429   
3                                 20.100000   
4                                 20.100000   
5                                       NaN   
6                                 22.950000   
7                                 22.727000   
8                                  4.269551   

   February Average Temperature (Midwest)  \
0                               34.300000   
1                               33.853333   
2                                3.987994   
3                               29.700000   
4                               29.700000   
5                                     NaN   
6                               32.950000   
7                               32.799000   
8                                4.807664   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                    27.400000                  42.800000   
1                                    27.293333                  42.966000   
2                                     4.871853                   2.918167   
3                                    23.200000                  41.310000   
4                                    23.200000                  41.310000   
5                                          NaN                        NaN   
6                                    26.950000                  41.810000   
7                                    26.631000                  41.669800   
8                                     4.453819                   2.941433   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                              33.000000                            43.500000   
1                              32.966667                            43.026667   
2                               3.003490                             4.148402   
3                              35.700000                            44.100000   
4                              35.700000                            44.100000   
5                                    NaN                                  NaN   
6                              32.400000                            42.700000   
7                              32.369000                            42.620000   
8                               3.994534                             4.649862   

   March Average Temperature (Pennsylvania)  
0                                 37.200000  
1                                 36.320000  
2                                  3.550091  
3                                 37.800000  
4                                 37.800000  
5                                       NaN  
6                                 35.600000  
7                                 35.947000  
8                                  4.333810

In [123]:
print(no_shadow.shape[0])
print(partial_shadow.shape[0])
print(full_shadow.shape[0])

15
1
100


In [130]:
df.loc[df['Shadow'] == 0]

Unnamed: 0  Year Punxsutawney Phil  February Average Temperature  \
9            9  1895         No Record                         26.60   
10          10  1896         No Record                         35.04   
11          11  1897         No Record                         33.39   
13          13  1899         No Record                         25.50   
16          16  1902         No Record                         31.46   
57          57  1943         No Record                         36.28   

    February Average Temperature (Northeast)  \
9                                       15.6   
10                                      22.2   
11                                      23.6   
13                                      18.1   
16                                      20.1   
57                                      23.9   

    February Average Temperature (Midwest)  \
9                                     21.9   
10                                    33.5   
11                                    34.7   
13                                    22.2   
16                                    23.6   
57                                    35.4   

    February Average Temperature (Pennsylvania)  March Average Temperature  \
9                                          17.0                      39.97   
10                                         26.6                      38.03   
11                                         27.9                      38.79   
13                                         20.0                      37.63   
16                                         21.0                      41.58   
57                                         27.6                      38.50   

    March Average Temperature (Northeast)  \
9                                    27.6   
10                                   25.3   
11                                   32.0   
13                                   29.3   
16                                   37.1   
57                                   30.5   

    March Average Temperature (Midwest)  \
9                                  40.2   
10                                 36.9   
11                                 44.0   
13                                 38.4   
16                                 43.9   
57                                 38.5   

    March Average Temperature (Pennsylvania)  Shadow  
9                                       31.3     0.0  
10                                      27.8     0.0  
11                                      36.9     0.0  
13                                      34.0     0.0  
16                                      38.8     0.0  
57                                      34.7     0.0

In [131]:
df.loc[9, 'Punxsutawney Phil'] = 'Full Shadow'
df.loc[9, 'Shadow'] = 3
df.loc[10, 'Punxsutawney Phil'] = 'Full Shadow'
df.loc[10, 'Shadow'] = 3
df.loc[11, 'Punxsutawney Phil'] = 'No Shadow'
df.loc[11, 'Shadow'] = 1
df.loc[13, 'Punxsutawney Phil'] = 'Full Shadow'
df.loc[13, 'Shadow'] = 3
df.loc[16, 'Punxsutawney Phil'] = 'Partial Shadow'
df.loc[16, 'Shadow'] = 2
df.loc[57, 'Punxsutawney Phil'] = 'Full Shadow'
df.loc[57, 'Shadow'] = 3

In [132]:
df

Unnamed: 0  Year Punxsutawney Phil  February Average Temperature  \
9             9  1895       Full Shadow                         26.60   
10           10  1896       Full Shadow                         35.04   
11           11  1897         No Shadow                         33.39   
12           12  1898       Full Shadow                         35.37   
13           13  1899       Full Shadow                         25.50   
..          ...   ...               ...                           ...   
126         126  2012       Full Shadow                         37.51   
127         127  2013         No Shadow                         34.77   
128         128  2014       Full Shadow                         32.13   
129         129  2015       Full Shadow                         32.99   
130         130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
9                                        15.6   
10                                       22.2   
11                                       23.6   
12                                       24.8   
13                                       18.1   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
9                                      21.9   
10                                     33.5   
11                                     34.7   
12                                     33.3   
13                                     22.2   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
9                                           17.0                      39.97   
10                                          26.6                      38.03   
11                                          27.9                      38.79   
12                                          26.7                      41.05   
13                                          20.0                      37.63   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
9                                     27.6   
10                                    25.3   
11                                    32.0   
12                                    38.0   
13                                    29.3   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
9                                   40.2   
10                                  36.9   
11                                  44.0   
12                                  46.0   
13                                  38.4   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                             

The method I used in this question was comparing the temperature values for each year that had 'No Record' with the median, mean, and standard deviation of the temperature values for the different possible values of 'Shadow' individually. I also took into account the frequency of each value as 'Partial Shadow' only occured once, 'No Shadow' occured 15 times, and 'Full Shadow' occured 100 times. I compared each of the 8 temperature values and saw which of the 3 types it matched with the most.

#### Question 7 - If we are trying to write a machine learning program which predicts the type of shadow for the year, which column(s) should be the independent variables X? Why? 

#### Use pandas to assign a variable X to be all of the rows of the data set with the column(s) that represent the independent variables.

'Year', 'February Average Temperature', 'February Average Temperature (Northeast)', 'February Average Temperature (Midwest)', 'February Average Temperature (Pennsylvania)', 'March Average Temperature', 'March Average Temperature (Northeast)', 'March Average Temperature (Midwest)', and 'March Average Temperature (Pennsylvania)' should all be independent variables as they can all be used in order to predict the shadow for a given year and do not store the actual information about the shadow themselves.

In [135]:
X = df
X = X.drop(['Unnamed: 0', 'Punxsutawney Phil', 'Shadow'], axis = 1)

In [136]:
X

Year  February Average Temperature  \
9    1895                         26.60   
10   1896                         35.04   
11   1897                         33.39   
12   1898                         35.37   
13   1899                         25.50   
..    ...                           ...   
126  2012                         37.51   
127  2013                         34.77   
128  2014                         32.13   
129  2015                         32.99   
130  2016                         39.47   

     February Average Temperature (Northeast)  \
9                                        15.6   
10                                       22.2   
11                                       23.6   
12                                       24.8   
13                                       18.1   
..                                        ...   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   

     February Average Temperature (Midwest)  \
9                                      21.9   
10                                     33.5   
11                                     34.7   
12                                     33.3   
13                                     22.2   
..                                      ...   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
9                                           17.0                      39.97   
10                                          26.6                      38.03   
11                                          27.9                      38.79   
12                                          26.7                      41.05   
13                                          20.0                      37.63   
..                                           ...                        ...   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
9                                     27.6   
10                                    25.3   
11                                    32.0   
12                                    38.0   
13                                    29.3   
..                                     ...   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   

     March Average Temperature (Midwest)  \
9                                   40.2   
10                                  36.9   
11                                  44.0   
12                                  46.0   
13                                  38.4   
..                                   ...   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                                 41.6   
130                                 49.5   

     March Average Temperature (Pennsylvania)  
9                                        31.3  
10                                       27.8  
11                                       36.9  
12                                       42.0  
13                                       34.0  
..             

#### Question 8 - If we are trying to write a machine learning program which predicts the type of shadow for the year, which column(s) should be the dependent variables Y? Why?  

#### Use pandas to assign a variable Y to be all of the rows of the data set with the column(s) that represent the dependent variables.

'Punxsutawney Phil' and 'Shadow' should be the dependent variable as they store the type of shadow which is the information that we are trying to predict.

In [137]:
Y = df
Y = Y.drop(['Unnamed: 0', 'Year', 'February Average Temperature', 'February Average Temperature (Northeast)', 'February Average Temperature (Midwest)', 'February Average Temperature (Pennsylvania)', 'March Average Temperature', 'March Average Temperature (Northeast)', 'March Average Temperature (Midwest)', 'March Average Temperature (Pennsylvania)'], axis = 1)

In [138]:
Y

Punxsutawney Phil  Shadow
9         Full Shadow     3.0
10        Full Shadow     3.0
11          No Shadow     1.0
12        Full Shadow     3.0
13        Full Shadow     3.0
..                ...     ...
126       Full Shadow     3.0
127         No Shadow     1.0
128       Full Shadow     3.0
129       Full Shadow     3.0
130         No Shadow     1.0

[122 rows x 2 columns]

#### Question 9 - Using the train_test_split function from the sklearn module, use a 70/30 split validation to split the data set into train and test. If you follow the example at the following link: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

Note: It's common practice to use the following variable names for unpacking the result:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=(YOUR VALUE))

sklearn will create a proper split where the following statements hold true:
X_train is your train set independent variables
Y_train is your train set dependent variables
X_test is your test set independent variables
Y_test is your test set dependent variables

In [139]:
from sklearn.model_selection import train_test_split

In [140]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, train_size=0.7)

In [141]:
X_train

Year  February Average Temperature  \
70  1956                         32.27   
68  1954                         41.41   
10  1896                         35.04   
52  1938                         36.00   
16  1902                         31.46   
..   ...                           ...   
83  1969                         32.81   
67  1953                         35.87   
48  1934                         34.32   
94  1980                         32.85   
89  1975                         32.76   

    February Average Temperature (Northeast)  \
70                                      25.6   
68                                      30.0   
10                                      22.2   
52                                      26.0   
16                                      20.1   
..                                       ...   
83                                      23.7   
67                                      28.0   
48                                      10.4   
94                                      20.0   
89                                      24.7   

    February Average Temperature (Midwest)  \
70                                    36.1   
68                                    41.1   
10                                    33.5   
52                                    40.2   
16                                    23.6   
..                                     ...   
83                                    33.0   
67                                    37.4   
48                                    27.2   
94                                    26.6   
89                                    33.5   

    February Average Temperature (Pennsylvania)  March Average Temperature  \
70                                         29.8                      40.75   
68                                         33.6                      39.25   
10                                         26.6                      38.03   
52                                         31.4                      45.01   
16                                         21.0                      41.58   
..                                          ...                        ...   
83                                         26.7                      36.03   
67                                         31.8                      43.75   
48                                         15.2                      42.78   
94                                         23.0                      39.36   
89                                         28.8                      38.70   

    March Average Temperature (Northeast)  \
70                                   27.9   
68                                   32.2   
10                                   25.3   
52                                   35.0   
16                                   37.1   
..                                    ...   
83                                   30.1   
67                                   34.8   
48                                   29.2   
94                                   31.0   
89                                   30.4   

    March Average Temperature (Midwest)  \
70                                 42.1   
68                                 40.2   
10                                 36.9   
52                                 49.6   
16                                 43.9   
..                                  ...   
83                                 36.2   
67                                 44.5   
48                                 38.8   
94                                 38.5   
89                                 38.6   

    March Average Temperature (Pennsylvania)  
70                                      32.9  
68                                      35.0  
10                                      27.8  
52                                      40.1  
16                                      38.8  
..                                       ...  
83                                      32.7 

In [142]:
Y_train

Punxsutawney Phil  Shadow
70       Full Shadow     3.0
68       Full Shadow     3.0
10       Full Shadow     3.0
52       Full Shadow     3.0
16    Partial Shadow     2.0
..               ...     ...
83       Full Shadow     3.0
67       Full Shadow     3.0
48         No Shadow     1.0
94       Full Shadow     3.0
89         No Shadow     1.0

[85 rows x 2 columns]

In [143]:
X_test

Year  February Average Temperature  \
114  2000                         39.78   
9    1895                         26.60   
27   1913                         29.50   
129  2015                         32.99   
29   1915                         35.94   
23   1909                         35.02   
56   1942                         30.74   
87   1973                         33.58   
93   1979                         28.13   
64   1950                         35.55   
126  2012                         37.51   
69   1955                         31.05   
97   1983                         36.86   
77   1963                         34.12   
32   1918                         33.82   
33   1919                         32.25   
105  1991                         39.81   
71   1957                         38.07   
26   1912                         31.08   
53   1939                         29.98   
122  2008                         34.70   
58   1944                         34.50   
90   1976                         39.56   
130  2016                         39.47   
49   1935                         36.01   
19   1905                         26.94   
104  1990                         36.70   
65   1951                         34.54   
88   1974                         34.74   
96   1982                         32.59   
37   1923                         29.57   
47   1933                         29.59   
82   1968                         32.92   
20   1906                         33.69   
25   1911                         33.66   
92   1978                         27.99   
101  1987                         37.38   

     February Average Temperature (Northeast)  \
114                                      27.2   
9                                        15.6   
27                                       20.2   
129                                      12.1   
29                                       26.7   
23                                       26.4   
56                                       20.1   
87                                       21.8   
93                                       13.9   
64                                       21.6   
126                                      30.6   
69                                       24.2   
97                                       26.2   
77                                       16.2   
32                                       19.8   
33                                       25.4   
105                                      28.7   
71                                       27.3   
26                                       18.6   
53                                       24.9   
122                                      25.4   
58                                       22.4   
90                                       28.8   
130                                      28.1   
49                                       21.1   
19                                       15.2   
104                                      28.0   
65                                       26.0   
88                                       21.7   
96                                       22.9   
37                                       16.2   
47                                       25.8   
82                                       18.4   
20                                       22.0   
25                                       21.4   
92                                       15.6   
101                                      21.8   

     February Average Temperature (Midwest)  \
114                                    40.1   
9                                      21.9   
27                                     29.6   
129                                    23.4   
29                                     37.6   
23                                     37.2   
56                                     29.7   
87                                     32.7   
93                                     22.9   
64                                   

In [144]:
Y_test

Punxsutawney Phil  Shadow
114       Full Shadow     3.0
9         Full Shadow     3.0
27        Full Shadow     3.0
129       Full Shadow     3.0
29        Full Shadow     3.0
23        Full Shadow     3.0
56     Partial Shadow     2.0
87        Full Shadow     3.0
93        Full Shadow     3.0
64          No Shadow     1.0
126       Full Shadow     3.0
69        Full Shadow     3.0
97          No Shadow     1.0
77        Full Shadow     3.0
32        Full Shadow     3.0
33        Full Shadow     3.0
105       Full Shadow     3.0
71        Full Shadow     3.0
26        Full Shadow     3.0
53        Full Shadow     3.0
122       Full Shadow     3.0
58        Full Shadow     3.0
90        Full Shadow     3.0
130         No Shadow     1.0
49        Full Shadow     3.0
19        Full Shadow     3.0
104         No Shadow     1.0
65        Full Shadow     3.0
88        Full Shadow     3.0
96        Full Shadow     3.0
37        Full Shadow     3.0
47        Full Shadow     3.0
82        Full Shadow     3.0
20        Full Shadow     3.0
25        Full Shadow     3.0
92        Full Shadow     3.0
101       Full Shadow     3.0

#### Question 10 - Should the number of rows in X_train == number of rows in Y_train? Also, should the number of rows in X_test == Y_test? Why or why not? Validate your answer with code below:

The number of rows in X_train should equal the number of rows in Y_train and the number of rows in X_test should equal the number of rows in Y_test. This is because both of the train datasets have to refer to all the same rows in order to accurately train the machine learning program. The same goes for the test datasets in order to accurate test the program. This is also why the train_test_split function symmetrically splits the rows in X and Y.

In [145]:
print(X_train.shape[0] == Y_train.shape[0])

True


In [146]:
print(X_test.shape[0] == Y_test.shape[0])

True
